File: dmm.pl
#!/usr/bin/perl
use strict; use utf8;
use JSON;
use LWP::Simple;
use XML::RSS::Parser;

binmode(STDOUT,":utf8");

my $DMM = "http://www.dmm.co.jp";
my @DOM = ( "digital/videoa/-", "mono/dvd/-" );
my @RSS = ( "rss/=", "=/rss=create" );

my @PAGES_URL; 
foreach ( qw(0 1) ){ $PAGES_URL[$_] = "$DMM/$DOM[$_]/list/$RSS[$_]/article=maker/sort=release_date"; }

my @MORAS = qw(ya yu yo wa wo nn);
foreach my $c ('',qw(k s t n h m r)){ foreach (qw(a i u e o)) { push(@MORAS,$c.$_); } }

my $content;
my $p = XML::RSS::Parser->new;

sub write_to_file
{
	my $fn = shift; open ( DMM, ">$fn" ) or die "Can't open: $!\n";
	print DMM JSON->new->utf8->encode( @_ ); close( DMM );
}

sub read_from_file
{
	my $fn = shift; 
	my $_text = do { open( DMM, "<:encoding(UTF-8)", $fn) or die("Can't open: $!\n"); local $/; <DMM> };
	return JSON->new->decode($_text);
}

sub get_count
{
	my ( $type, $id, $domain ) = @_ ;
	my $ct; my @count = (0,0);
	  
	if ( $domain & 1 )
	{
		$ct = get( "$DMM/$DOM[0]/list/=/article=$type/id=$id" );
		($count[0]) = $ct =~ m/(\d+).*タイトル中/ig ;
	}

	if ( $domain & 2 )
	{
		$ct = get( "$DMM/$DOM[1]/list/=/article=$type/id=$id" );
		($count[1]) = $ct =~ m/(\d+).*?タイトル中/ ;
	}

	# if ( $domain == 3 ){ return @count; } else { return $count[$domain-1]; }
	return @count;
}

sub get_genres
{
	my %genres;

	$content = get( "$DMM/$DOM[0]/genre/" );
	foreach ( $content =~ m/article=keyword(.+?)<\/a>/mg )
	{
		my @g = $_ =~ m/id=(\d+).">(.+)/m; chomp $g[1];
		if ( $g[0] ne '' and $g[1] ne '' ){ $genres{$g[0]} = $g[1]; }
	}

	$content = get( "$DMM/$DOM[1]/genre/" );
	foreach ( $content =~ m/article=keyword(.+?)<\/a>/mg )
	{
		my @g = $_ =~ m/id=(\d+).">(.+)/m; chomp $g[1];
		if ( $g[0] ne '' && ( ! $genres{$g[0]} ) ){ $genres{$g[0]} = $g[1]; }
	}

	return \%genres;
}

sub get_makers
{
	my %makers;

	foreach ( @MORAS )
	{

		$content = get( "$DMM/$DOM[0]/maker/=/keyword=$_/" );

		foreach ( $content =~ m/(?s)class="d-unit">(.+?)<!--\/d-unit-->/mg )
		{
			my @m = $_ =~ m/(?s)id=(\d+).+?src="(.+?)".+?llarge">(.+?)<.+?\n(.*)<\/div>/m;

			my $id = shift(@m); if ( $id eq '' ){ next; }

			if ( $m[2] ne '' ){ $m[2] =~ s/.*<p>(.+?)<\/p>.*/\1/; }
			if ( $m[0] =~ m/noimage/ ){ $m[0] = ''; }

			$m[3] = 1;

			my %maker; @maker{qw(img name description domain)} = @m;
			$makers{$id} = \%maker;

		}

		$content = get( "$DMM/$DOM[1]/maker/=/keyword=$_/" );

		foreach ( $content =~ m/(?s)class="w50">(.+?)class="clear"/mg )
		{
			my @m = $_ =~ m/(?s)id=(\d+).+?src="(.+?)".+?bold">(.+?)<.+?\n(.*)<\/div>.*/m;

			my $id = shift(@m); if ( $id eq '' ){ next; }

			if ( $makers{$id} ){ $makers{$id}{"domain"} = 3; next; }
			
			if ( $m[2] ne '' ){ $m[2] =~ s/.*<p>(.+?)<\/p>.*/\1/; }
			if ( $m[0] =~ m/noimage/ ){ $m[0] = ''; }

			$m[3] = 2;

			my %maker; @maker{qw(img name description domain)} = @m;
			$makers{$id} = \%maker;
		}
	}

	return \%makers;
}

sub get_actresses
{
	my %acts;

	foreach ( @MORAS )
	{
		$content = get( "$DMM/$DOM[0]/actress/=/keyword=$_/" );

		foreach ( $content =~ m/article=actress(.+?)<\/a>/mg )
		{
			my @a = $_ =~ m/id=(\d+).+?src="(.+?)".+?"><br>(.+)/m;

			my $id = shift(@a); if ( $id eq '' ){ next; }

			if ( $a[0] =~ m/noimage/ ){ $a[0] = ''; }

			$a[2] = 1;

			my %actress; @actress{qw(img name domain)} = @a;
			$acts{$id} = \%actress;
		}

		$content = get( "$DMM/$DOM[1]/actress/=/keyword=$_/" );

		foreach ( $content =~ m/article=actress(.+?)<\/a>/mg )
		{
			my @a = $_ =~ m/id=(\d+).+?src="(.+?)".+?"><br>(.+)/m;

			my $id = shift(@a); if ( $id eq '' ){ next; }

			if ( $acts{$id} ){ $acts{$id}{"domain"} = 3; next; }

			if ( $a[0] =~ m/noimage/ ){ $a[0] = ''; }

			$a[2] = 2;

			my %actress; @actress{qw(img name domain)} = @a;
			$acts{$id} = \%actress;
		}
	}

	return \%acts;
}

sub update_counts
{
	my $makers = shift;
	my @update; my @updated;

	for my $id ( keys %{$makers} )
	{
		my @s_c = get_count( "maker", $id, $makers->{$id}{'domain'} );

		if ( exists($makers->{$id}{'count'}) )
		{
			if ( not ( @s_c ~~ @{$makers->{$id}{'count'}} ) )
			{
				print "$id updated:";
				foreach ( qw(0 1) )
				{
					$update[$_] = $s_c[$_] - @{$makers->{$id}{'count'}}[$_];
					if ( $update[$_] ne 0 ){ print "\t$_\t+$update[$_]"; }
				}
				push( @updated, $id );
				print "\n";
			}
		}
		else
		{
			print "$id initialized with ( $s_c[0], $s_c[1] ).\n";
		}
		
		@{$makers->{$id}{'count'}} = @s_c;
	}

	return @updated;
}

sub get_pages
{
	my $feed;
	my ( $works, $q, $i_start, $i_end, $step ) = @_ ;

	my $p_end = int( $i_end/$step );
	if ( $i_end%$step > 0 ){ $p_end += 1; }

	# print "$i_start\t$i_end\t$p_end\t$step\n";

	for ( my $i = int( $i_start/$step )+1; $i <= $p_end; $i++ )
	{
		$feed = $p->parse_uri( "$q/limit=$step/page=$i/" );

		if ( not defined $feed )
		{
			if ( $step != 1 )
			{
				get_pages( $works, $q, ($i-1)*$step+1, $i*$step, $step/5 );
			}
			else { print "error at $i\n"; }
		}
		else
		{
			foreach ( $feed->query('//item') ){ push( @{$works}, get_items($_) ); }
		}
	}
}

sub get_items
{
	my %v; my $content = $_->query('content:encoded')->text_content;

	$v{'title'} = $_->query('title')->text_content;
	$v{'url'} = $_->query('link')->text_content;
	$v{'description'} = $_->query('description')->text_content;
	$v{'date'} = $_->query('dc:date')->text_content;

	($v{'cid'}) = $v{'url'} =~ m/cid=([^\/]+)/ ;
	($v{'runtime'}) = $content =~ m/>([0-9]+)分/ ;
	($v{'label'}) = $content =~ m/article=label\/id=([0-9]+)/g ;
	($v{'series'}) = $content =~ m/article=series\/id=([0-9]+)/g ;
	$v{'actress'} = [ $content =~ m/article=actress\/id=([0-9]+)/g ];

	$v{'tags'} = [ $content =~ m/article=keyword\/id=([0-9]+)/g ];

	#print "$_ $v{$_}\n" for keys %v; print "\n";
	#print "$v{'cid'}\t$v{'title'}\n";
	
	return \%v;
}

sub get_works
{
	my $makers = shift;

	for my $id ( keys %{$makers} )
	{
 		my @works = ();

		print "$id: ";

		foreach ( qw(0 1) )
		{
			get_pages( \@works, "$PAGES_URL[$_]/id=$id", 1, @{$makers->{$id}{'count'}}[$_], 125 );
		}

		# foreach ( @works ) { print "$_ $work->{$_}\t" for ( keys %{$work} ); print "\n"; }

		write_to_file( "works/$id.json", \@works );
		
		print @works." works written.\n";
	}
}

#my %keywords = %{&get_genres}; write_to_file("genres.json", \%keywords );

$|=1;

my %studios; my %st_updated;

if ( -e "makers.json" )
{
	print "Reading from file...";
	%studios = %{&read_from_file("makers.json")};
	print " OK.\n";

	print "Getting online version...";
	%st_updated = %{&get_makers};
	print " OK.\n";

	if ( %studios != %st_updated )
	{
		print "Maker list updated\n";

		# update maker list
	}
	else
	{
		print "Local database matches online. Continuing.\n";
	}
}
else
{
	print "Database not found. Getting online version...";
	%studios = %{&get_makers};
	print " OK.\n";
}

print "Updating counts...\n";
my @updated = update_counts(\%studios);
print " OK.\n";
%st_updated = map { $_ => $studios{$_} } @updated;
print @updated." makers require update. Updating...\n";
get_works(\%st_updated); 
print " OK.\n";
write_to_file("makers.json", \%studios);

File: DMM.py
import re
import requests
from bs4 import BeautifulSoup
from datetime import timedelta

class DMM:
    """Website as an object."""

    REALMS = ( 'digital/videoa', 'mono/dvd' )

    MORAS = [ c.strip()+v for c in ' kstnhmr' for v in 'aiueo' ]
    MORAS.extend(['ya','yu','yo','wa','wo','nn'])

    L_PAGE = 'list-boxcaptside list-boxpagenation'

    ART_ID = re.compile(r'article=(\w+)/id=(\d+)')

    def normalize( self, t ):
        """Standardize characters over realms."""
        return re.sub('〜','～',t)

    def get_soup( self, page, realm=None ):
        """Get page as a BeautifulSoup."""

        if realm is not None: page = '%s/-/%s' % ( self.REALMS[realm], page )
        r = requests.get( 'http://www.dmm.co.jp/%s' % page )

        return BeautifulSoup( r.text, 'html.parser' )

    def get_id( self, a ):
        """Get ID from <a> tag href."""
        rid = re.compile(r'/id=(\d+)')
        try:
            return int( rid.search(a.get('href')).group(1) )
        except AttributeError:
            return None

    def get_filename( self, img ):
        """Get filename from <img> tag src."""
        name = img.get('src').rsplit('/',1)[-1].split('.')
        if name[0] == 'noimage':
            return ''
        else:
            return name[0]

    def get_article( self, realm, article, a_id ):
        """Get info of given ID."""
        item = {}
        soup = self.get_soup( "list/=/article=%s/id=%s/" % ( article, a_id ), realm )

        pg = soup.find('div',class_=self.L_PAGE)

        item['count'] = 0
        if pg:
            numz = pg.p.string.replace(',','')
            item['count'] = int(re.match(r'(\d+)',numz).group(1))

        item['name'] = re.sub( r' -[^-]+- DMM.R18$', '', soup.title.string )

        if article == 'actress' or article == 'director':
            ft = re.split( r'[()（）]+', item['name'] )[:-1]

            if ft:
                item['name'] = ft[0]
                item['furi'] = ft[-1]

                if len(ft) == 3:
                    item['alias'] = ft[1]

        return item

    def get_work_page( self, realm, cid ):
        """Get info page of given CID."""

        work = {}
        soup = self.get_soup( "detail/=/cid=%s/" % cid, realm )

        detail = soup.find('div',class_='page-detail').table

        sample_imgs = detail.find('div',id='sample-image-block')
        if sample_imgs: work['sample_images'] = len( sample_imgs('a') )

        for cell in detail.table('td'):
            try:
                d = self.ART_ID.search( cell.a.get('href') ).groups()
            except AttributeError:
                dt = re.match(r'(\d+)/(\d+)/(\d+)',next(cell.stripped_strings,''))
                if 'date' not in work and dt: work['date'] = '-'.join(dt.groups())
                continue
            if d[0] == 'director': work['director'] = d[1]

        return work

    def get_related( self, realm, cid ):
        """Get related CIDs and their realms."""
        cds = re.compile(r'dmm.co.jp/(.+)/-/detail/=/cid=(\w+)')
        path = "misc/-/mutual-link/ajax-index/=/cid={0}/service={1[0]}/shop={1[1]}/"

        related = []
        soup = self.get_soup( path.format( cid, self.REALMS[realm].split('/') ) )

        for l in soup('li'):
            r = cds.search( l.a.get('href') )
            if r.group(1) in ( 'mono/dvd', 'digital/videoa' ): related.append( r.groups() )

        return related

    def get_image_path( self, realm, pid, param='pt' ):
        """Get image path."""
        IMG_REALM = ( 'digital/video', 'mono/movie/adult' )

        if param.startswith('jp'): realm = 0

        return "{0}/{1}/{1}{2}.jpg".format( IMG_REALM[realm], pid, param )

    def get_sample_vid_path( self, cid, param='sm' ):
        """Get sample video path."""

        def get_sample_vid_id( cid ):
            flv = re.compile(r'flashvars.(\w+) = "?(\w+)"?')

            soup = self.get_soup( "service/-/flash/=/cid=%s/" % cid )

            for s in soup.find_all('script',string=True):
                for fv in flv.finditer(s.string):
                    if fv.group(1) == 'cid': return fv.group(2)

        # sizes = ( 'sm', 'dm', 'dmb' )

        vid = get_sample_vid_id( cid )

        if not vid: return ''

        return "litevideo/freepv/{0:.1}/{0:.3}/{0}/{0}_{1}_w.mp4".format( vid, param )

    def identify_maker( self, pid ):

        makers = {
            '1'  : True, '13' : True, '53' : 40039, '59' : True, '61' : 40047, '84' : 40071,
            '118': True, '171': True, '172': 40185, '422': True, '433': True, '436': 45061,
            'h_068': True, 'h_094': True, 'h_244': True, 'h_254': True, 'h_259': True,
            'h_422': 45667, 'h_565': True, 'h_606': True, 'h_796': True, 'h_843': True,
        }

        if not parts[0]:
            maker = True
            if parts[1] == 'bnsps': maker = 45249
        elif parts[0].startswith('55'):
            maker = 40041
        elif parts[0] in makers:
            maker = makers[parts[0]]

        if not maker:
            return None

    def rename( self, pid, maker ):
        """Get DVD name from pid and maker."""

        def get_num(p,digits=3): return '-{:0{d}d}'.format(int(p[1]),d=digits)

        def get_suffix(p):
            if not p[2]: return ''
            return p[2] if '_' in p[2] else '-%s' % p[2]

        def get_txt(p,digits): return p[0] + get_num(p,digits) + get_suffix(p)

        id_base = ( r'^(?:h_)?(?:\d+)?', r'((?:d1)?[a-z]+(?:3d)?[a-z]*)', r'(\d+)([a-z]+|_\d)?$' )

        default_parser = { 're': ''.join(id_base), 'digits': 3, 'txt': get_txt }

        parsers = {
            1398 : { 're': r'^(d1clymax|dcb1|[a-z]+)' + id_base[2] },
            40039 : { 'digits': 4 }, 45667 : { 'digits': 4 },
            40041 : { 're': r'^(?:55|57)(t28|\d*[a-z]+)(\d+)([a-z])?$' },
            45249 : { 'massage': lambda p: ('nsps',p[1]) if p[0] == 'bnsps' else p }
        }

        try:
            parser = parsers[int(maker)]
            for k in ( 're', 'digits', 'txt' ):
                if k not in parser: parser[k] = default_parser[k]
        except KeyError:
            parser = default_parser
        except ( TypeError, ValueError ):
            print( "Error: Could not get maker %s" % maker )
            return None

        try:
            parts = re.match(parser['re'],pid).groups()
        except AttributeError:
            print( "Error: Could not identify a base for %s" % pid )
            return None

        if 'massage' in parser: parts = parser['massage']( parts )

        return parser['txt']( parts, parser['digits'] ).upper()

    def get_keywords( self ):
        """Get keywords in array."""

        def tag(link, category):
            cats = ('シチュエーション','ＡＶ女優タイプ','コスチューム','ジャンル','プレイ','その他')

            _id = self.get_id(link)
            if not _id: return None

            return { '_id': _id, 'name': link.string, 'category': cats.index(category) }

        tags = []

        soup = self.get_soup( 'genre/', 0 )

        for section in soup.find_all('div', id=re.compile('^list')):
            category = section.div.string
            if category == "おすすめジャンル": continue
            if category == "タイプ": category = "ＡＶ女優タイプ"

            tags.extend([ tag(link,category) for link in section.ul('a') ])

        soup = self.get_soup( 'genre/', 1 )

        for section in soup.find_all('table', class_='sect02'):
            category = section.get('summary')
            tags.extend([ tag(link,category) for link in section('a') ])

        return list(filter(None.__ne__, tags))

    def get_makers( self, mora, callback=print ):

        def maker( item, realm ):
            m = {}

            m['_id'] = self.get_id(item.a)
            if not m['_id']: return None

            m['roma'] = self.get_filename(item.img)

            if realm == 0:
                m['name'] = item.find(class_='d-ttllarge').string
                try:
                    m['description'] = item.p.string.strip()
                except AttributeError:
                    pass

            elif realm == 1:
                m['name'] = item.img.get('alt')
                m['description'] = self.normalize(item.br.string.strip())

            return m

        search = 'maker/=/keyword=%s/' % mora 

        soup = self.get_soup( search, 0 )

        for div in soup.find_all('div', class_='d-boxpicdata d-smalltmb'):
            callback( maker( div, 0 ) )

        soup = self.get_soup( search, 1 )

        for cell in soup.find_all('td',class_='w50'):
            callback( maker( cell, 1 ) )

        extra_base = soup.find(class_='list-table mg-t12')
        if extra_base:
            for link in extra_base('a'):
                _id = self.get_id(link)
                if not _id: continue
                callback({ '_id': _id, 'name': link.string })

    def get_works( self, realm, m_id, count, page=1, callback=print ):

        def get_rss( s, r, q ):

            URL = 'http://www.dmm.co.jp/{0}/-/list/{1}/article=maker/sort=release_date/{2}/' 
            RSS = ( "rss/=", "=/rss=create" )

            print( "%s %s:" % (r, q), end="\t" )
            r = s.get( URL.format( self.REALMS[r], RSS[r], q ) )
            print( r.elapsed, end="\r" )

            r.encoding = 'utf-8'
            return BeautifulSoup( r.text, 'xml' )

        def parse_work_rss( soup ):

            properties = ( 'title', 'description', 'link', 'package', 'date' )

            work = { 'keywords': [], 'actresses': [] }

            for p in properties: work[p] = soup.find(p).string

            work['cid'] = re.search(r'cid=(\w+)', work['link']).group(1)
            work['pid'] = re.search(r'/(video|adult)/(\w+)', work['package']).group(2)
            work['released_date'] = work['date'].split('T')[0]

            content = BeautifulSoup( soup.encoded.string, 'html.parser' )

            for info in content('strong'):
                if '分' in info.next_sibling:
                    dur_mins = int( info.next_sibling.strip('分') )
                    work['runtime'] = timedelta( minutes = dur_mins )
                    break

            for link in content('a'):
                l = self.ART_ID.search( link.get('href') )
                if not l: continue
                if l.group(1) == 'keyword':
                    work['keywords'].append( l.group(2) )
                elif l.group(1) == 'actress':
                    work['actresses'].append( l.group(2) )
                else:
                    work[l.group(1)] = l.group(2)

            work['_id'] = self.rename( work['pid'], work['maker'] )
            work['realm'] = realm

            return work

        works = []
        worksession = requests.Session()

        while len(works) < count:
            soup = get_rss( worksession, realm, "id=%d/page=%d" % ( m_id, page ) )
            for item in soup('item'): works.append( callback( parse_work_rss(item) ) )
            page += 1

        return works

    def search_dmm( self, terms ):

        soup = self.get_soup( 'search/=/searchstr=%s/' % '%20'.join(terms) )

        for a in soup.find('div',class_='d-item')('a'):
            link = a.get('href') 

            for d in ('digital','mono'):
                if d not in link: continue
                try:
                    cid = re.search(r'cid=(\w+)', link).group(1)
                except AttributeError:
                    continue

                for t in terms:
                    if t not in cid: break
                else:
                    print( cid )

if __name__ == "__main__":

    dmm = DMM()
    # print( dmm.get_keywords() )
